stan/math/rev/core/vari.hpp

```diff
@@ -184,6 +186,400 @@ class vari_value<T, std::enable_if_t<std::is_floating_point<T>::value>>
 // For backwards compatability the default is double
 using vari = vari_value<double>;
 
+/**
+ * The variable implementation for Eigen dense matrix types.
+ *
+ * This class is complete (not abstract) and may be used for
+ * constants.
+ *
+ * A variable implementation is constructed with a constant
+ * value. It also stores the adjoint for storing the partial
+ * derivative with respect to the root of the derivative tree.
+ *
+ */
+template <typename T>
+class vari_value<T, std::enable_if_t<is_eigen_dense_base<T>::value>>
+    : public vari_base {
+ public:
+  /**
+   * `PlainObject` represents a user constructible type such as Matrix or Array
+   */
+  using PlainObject = std::decay_t<plain_type_t<T>>;
+  using eigen_scalar = value_type_t<PlainObject>;  // A floating point type
+  eigen_scalar* val_mem_;  // Pointer to memory allocated on the stack for val_
+  eigen_scalar* adj_mem_;  // Pointer to memory allocated on the stack for adj_
+  using Scalar = PlainObject;  // The underlying type for this class
+  using value_type = Scalar;   // The underlying type for this class
+  using eigen_map = Eigen::Map<PlainObject>;  // Maps for adj_ and val_
+  const Eigen::Index rows_;                   // Number of rows
+  const Eigen::Index cols_;                   // Number of cols
+  const Eigen::Index size_;                   // Size of rows * cols
```

Comment:
Can these all be functions that access the maps? Seems like they take up unnecessary memory.

Reply:
Good call!